File: backend/app/schemas/vocab_lab.py
import uuid

from pydantic import BaseModel


class UnitVocabStatsRead(BaseModel):
    unit_id: uuid.UUID
    unit_number: str
    title_ko: str | None
    title_en: str | None
    word_count: int
    mastered_count: int
    learning_count: int
    needs_review_count: int


class VocabWordRead(BaseModel):
    id: uuid.UUID
    korean: str
    english: str
    romanization: str | None
    notes: str | None
    image_url: str | None
    attempt_count: int
    correct_count: int
    mastery_score: float
    status: str
    last_attempt_at: str | None
    is_teacher_added: bool = False


class VocabQuestionRead(BaseModel):
    vocabulary_id: uuid.UUID
    question_type: str
    prompt: str
    options: list[str] | None
    image_url: str | None


class FlashcardReviewRequest(BaseModel):
    rating: str  # "again" | "hard" | "good" | "easy"


class FlashcardReviewResult(BaseModel):
    status: str
    mastery_score: float


class VocabAttemptRequest(BaseModel):
    question_type: str
    submitted_answer: str
    ui_mode: str | None = None


class VocabAttemptResult(BaseModel):
    is_correct: bool
    correct_answer: str


class QuickRecallResultRequest(BaseModel):
    total: int
    correct: int
    missed_vocabulary_ids: list[uuid.UUID]


class VocabSearchResultRead(BaseModel):
    id: uuid.UUID
    korean: str
    english: str
    romanization: str | None
    notes: str | None
    image_url: str | None
    unit_number: str
    unit_title_ko: str | None
    is_teacher_added: bool = False


class NewVocabularyItemRead(BaseModel):
    id: uuid.UUID
    korean: str
    english: str
    romanization: str | None
    notes: str | None
    image_url: str | None
    unit_number: str
    unit_title_ko: str | None

File: frontend/src/types/vocabLab.ts
// Vocabulary Memory Lab type contracts - all data comes from the API,
// real canonical vocabulary + this user's own progress. Never hardcoded.

export interface UnitVocabStats {
  unit_id: string
  unit_number: string
  title_ko: string | null
  title_en: string | null
  word_count: number
  mastered_count: number
  learning_count: number
  needs_review_count: number
}

export interface VocabWord {
  id: string
  korean: string
  english: string
  romanization: string | null
  notes: string | null
  image_url: string | null
  attempt_count: number
  correct_count: number
  mastery_score: number
  status: 'MASTERED' | 'LEARNING' | 'NEEDS_REVIEW'
  last_attempt_at: string | null
  is_teacher_added: boolean
}

export type QuestionType = 'ko_to_en_written' | 'en_to_ko_written' | 'multiple_choice' | 'context_to_word'

export interface VocabQuestion {
  vocabulary_id: string
  question_type: QuestionType
  prompt: string
  options: string[] | null
  image_url: string | null
}

export type FlashcardRating = 'again' | 'hard' | 'good' | 'easy'

export interface FlashcardReviewResult {
  status: string
  mastery_score: number
}

export interface VocabAttemptResult {
  is_correct: boolean
  correct_answer: string
}

export interface VocabSearchResult {
  id: string
  korean: string
  english: string
  romanization: string | null
  notes: string | null
  image_url: string | null
  unit_number: string
  unit_title_ko: string | null
  is_teacher_added: boolean
}

export interface NewVocabularyItem {
  id: string
  korean: string
  english: string
  romanization: string | null
  notes: string | null
  image_url: string | null
  unit_number: string
  unit_title_ko: string | null
}
